### PR TITLE
Main menu updating and pagination

### LINKED ISSUE #3212: [SDL 0268] Main Menu Updating and Pagination
Proposal: [Main Menu Updating and Pagination](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0268-main-menu-updating.md)

> This proposal seeks to add the ability to update the main menu with artworks and sub-menus on demand so that we can improve the performance and usability of the menu.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/883

### PR DESCRIPTION
**Pull request by JackLivio**
Fixes #3212 

This PR is **not ready** for review.

Still need to add unit test changes.

### Risk
This PR makes **minor** API changes.

### Summary
Implements proposal [0268-main-menu-updating](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0268-main-menu-updating.md)

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/interfaces/HMI_API.xml [hunk 1]
```diff
@@ -5663,6 +5678,33 @@
   </function>
   <function name="SendHapticData" messagetype="response">
   </function>
+
+  <function name="OnUpdateFile" messagetype="notification">
```
Comment (iCollin) [inline src/components/interfaces/HMI_API.xml:5682]: the proposal obviously meant notification, correct? we don't need to modify where the proposal says request?

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/on_ui_update_file_notification.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
+/*
+ Copyright (c) 2019, Ford Motor Company, Livio
```
Comment (iCollin) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/on_ui_update_file_notification.h:2]:
```suggestion
 Copyright (c) 2020, Ford Motor Company, Livio
```

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/on_ui_update_sub_menu_notification.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
+/*
+ Copyright (c) 2019, Ford Motor Company, Livio
```
Comment (iCollin) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/on_ui_update_sub_menu_notification.h:2]:
```suggestion
 Copyright (c) 2020, Ford Motor Company, Livio
```

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/on_update_file_notification.h [hunk 4]
```diff
@@ -0,0 +1,78 @@
+/*
+ Copyright (c) 2019, Ford Motor Company, Livio
```
Comment (iCollin) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/on_update_file_notification.h:2]:
```suggestion
 Copyright (c) 2020, Ford Motor Company, Livio
```

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/on_update_sub_menu_notification.h [hunk 5]
```diff
@@ -0,0 +1,79 @@
+/*
+ Copyright (c) 2019, Ford Motor Company, Livio
```
Comment (iCollin) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/on_update_sub_menu_notification.h:2]:
```suggestion
 Copyright (c) 2020, Ford Motor Company, Livio
```

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/on_ui_update_file_notification.cc [hunk 6]
```diff
@@ -0,0 +1,82 @@
+/*
+ Copyright (c) 2019, Ford Motor Company, Livio
```
Comment (iCollin) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/on_ui_update_file_notification.cc:2]:
```suggestion
 Copyright (c) 2020, Ford Motor Company, Livio
```

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/on_ui_update_sub_menu_notification.cc [hunk 7]
```diff
@@ -0,0 +1,81 @@
+/*
+ Copyright (c) 2019, Ford Motor Company, Livio
```
Comment (iCollin) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/on_ui_update_sub_menu_notification.cc:2]:
```suggestion
 Copyright (c) 2020, Ford Motor Company, Livio
```

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/on_update_file_notification.cc [hunk 8]
```diff
@@ -0,0 +1,62 @@
+/*
+ Copyright (c) 2019, Ford Motor Company, Livio
```
Comment (iCollin) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/on_update_file_notification.cc:2]:
```suggestion
 Copyright (c) 2020, Ford Motor Company, Livio
```

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/on_update_sub_menu_notification.cc [hunk 9]
```diff
@@ -0,0 +1,62 @@
+/*
+ Copyright (c) 2019, Ford Motor Company, Livio
```
Comment (iCollin) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/on_update_sub_menu_notification.cc:2]:
```suggestion
 Copyright (c) 2020, Ford Motor Company, Livio
```

File: src/components/interfaces/HMI_API.xml (hunk 1 above)
Comment (JackLivio) [reply src/components/interfaces/HMI_API.xml:5682]: Yes it should be notification.

File: src/components/utils/src/file_system.cc [hunk 10]
```diff
@@ -572,3 +572,8 @@ bool file_system::MoveFile(const std::string& src, const std::string& dst) {
   }
   return true;
 }
+
+std::string file_system::GetFileName(const std::string& full_path) {
+  fs::path p(full_path);
+  return p.filename().string();
+}
```
Comment (JackLivio) [inline src/components/utils/src/file_system.cc:579]:
```suggestion
}

```

**Pull request merged**